core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
-    //       indicates the metadata is not exist at least for now.
+    // we only need to check loading partition set but not leaving partition set here since there are three possible cases:
+    //    1) it is not in the leaving partitions set, hence safe to return COORDINATOR_LOAD_IN_PROGRESS
+    //    2) it is in the leaving partitions with a smaller epoch than the latest loading epoch, hence safe to return COORDINATOR_LOAD_IN_PROGRESS
+    //    3) it is in the leaving partition with a larger epoch, return true is also OK since the client will then retry
+    //       later be notified that this coordinator is no longer be the transaction coordinator for him
     //
-    //    4) it is NOT possible to be in the loading partition with the same epoch
-    if (leavingPartitions.exists(_.txnPartitionId == partitionId))
-      return None
-
-    transactionMetadataCache.get(partitionId).flatMap { cacheEntry =>
-      cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
-        case null => None
-        case txnMetadata => Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata))
+    //    4) it is NOT possible to be in the leaving partition with the same epoch
+    if (loadingPartitions.exists(_.txnPartitionId == partitionId))
+      return Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
+
+    transactionMetadataCache.get(partitionId) match {
+      case Some(cacheEntry) =>
+        cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
+          case null =>
+            Right(None)
+          case txnMetadata =>
+            // we only need to check leaving partition set but not loading partition set here since there are three possible cases:
+            //    1) it is not in the loading partitions set, hence safe to return NONE
+            //    2) it is in the loading partitions with a smaller epoch, hence safe to return NONE
+            //    3) it is in the loading partition with a larger epoch, return NONE is also fine as it
+            //       indicates the metadata is not exist at least for now.
+            //
+            //    4) it is NOT possible to be in the loading partition with the same epoch
+            if (leavingPartitions.exists(_.txnPartitionId == partitionId))
+              Right(None)
+            else
+              Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)))
+        }
+      case None =>
+        Left(Errors.NOT_COORDINATOR)
       }
-    }
   }
 
   /**
    * Add a new transaction metadata, or retrieve the metadata if it already exists with the associated transactional id
    * along with the current coordinator epoch for that belonging transaction topic partition
+   *
+   * Need to grab the state lock before calling this function
    */
   def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): CoordinatorEpochAndTxnMetadata = {
```

Comment:
Don't we need the read lock when `transactionMetadataCache.get` below? Also, why is this operation safe without checking the loading/leaving partitions but `getTransactionState` is not?